src/expressions.rs

```diff
@@ -30,3 +36,68 @@ pub(crate) fn user_full_name(user_expression: &gtk::Expression) -> gtk::Expressi
     })
     .upcast()
 }
+
+pub(crate) fn restriction_expression(chat: &Chat) -> gtk::Expression {
+    // chat type is (usually) constant so we don't need to update it
+    match chat.type_() {
+        ChatType::Supergroup(data) if !data.is_channel() => {
+            restriction_label_expression::<Supergroup, _>(data)
+        }
+        ChatType::BasicGroup(data) => restriction_label_expression::<BasicGroup, _>(data),
+        _ => gtk::ConstantExpression::new(&"").upcast(),
+    }
+}
+
+fn restriction_label_expression<T: glib::StaticType, V: glib::ToValue>(
+    value: &V,
+) -> gtk::Expression {
+    let member_status_expression = gtk::PropertyExpression::new(
+        T::static_type(),
+        Some(gtk::ConstantExpression::new(value)),
+        "status",
+    );
+    let permissions_expression = Chat::this_expression("permissions");
+
+    gtk::ClosureExpression::new::<String, _, _>(
+        &[member_status_expression, permissions_expression],
+        closure!(|_: glib::Object, status: BoxedChatMemberStatus, chat_permissions: BoxedChatPermissions| {
+            if chat_permissions.0.can_send_messages {
+                match status.0 {
+                ChatMemberStatus::Restricted(status) if !status.permissions.can_send_messages => {
+                    if status.restricted_until_date == 0 {
+                        gettext("The admins of this group have restricted you from writing here")
+                    } else {
+                        let date =
+                            glib::DateTime::from_unix_utc(status.restricted_until_date.into()).unwrap();
+
+                        gettext!(
+                            "The admins of this group have restricted you from writing here until {}",
+                            date.format(&if glib::DateTime::now_local()
+                                .unwrap()
+                                .difference(&date)
+                                .as_days()
+                                == 0
+                            {
+                                // Translators: This is a hours: minutes time format used in restriction label
+                                gettext("%l:%M %p")
+                            } else {
+                                // Translators: This is a full date format used in restriction label
+                                gettext("%B %e, %Y %l:%M %p")
+                            })
+                            .unwrap()
+                            .to_string()
+                        )
+                    }
+                }
+                _ => String::new(),
+            }
+        } else if !matches!(status.0, ChatMemberStatus::Creator(_) | ChatMemberStatus::Administrator(_)) {
+                            gettext("Writing messages isn't allowed in this group.")
+                        } else {
+                            String::new()
+                        }
+
```

Comment:
The formatting is wrong here, as rustfmt sometimes fails in macros. In such cases, I copy and paste the code to a location somewhere outside the macro with the same level of indentation, run rustfmt, and copy it back.

Reply:
Thanks for your tip, I'll fix it later